### PR TITLE
[WIP] Redstone support - Basics

### PR DESCRIPTION
**Pull request by kaenganxt**
As wanted in #381 i'm splitting my pull request.
## Features
- [x] Basic redstone support
- [x] Torch placement
- [x] Redstone torch support
- [x] Redstone dust support
## TODO
- Nothing
## Requirements
- None

---

_Edited by turt2live_

**Related Links:**
- Ticket: #116

### CODE REVIEW

**Comment by dequis** [comment]
This is 1200 lines vs 1500 of the original :(

**Comment by turt2live** [comment]
Your checklist as-is, assuming it works, is suitable for review. Fixing torches not burning out is kinda out of scope due to the size of this PR at this time.

**Comment by kaenganxt** [comment]
ok, i removed the WIP label. Then @turt2live i think it's ready for reviewing.

File: src/main/java/net/glowstone/RSManager.java [hunk 1]
```diff
@@ -0,0 +1,541 @@
...
+
+    /**
+     * The world this manager is assigned to.
+     */
+    private final GlowWorld world;
+
+    /**
+     * Various redstone-related sets, maps and lists.
+     */ 
+    private final Map<BlockVector, Integer> redPowerCurrent = new HashMap<>();
```
Comment (turt2live) [inline src/main/java/net/glowstone/RSManager.java:29]: I'm still not a fan of having this many collections. There must be a way to reduce the collections and still provide function without implementing useless code.

File: src/main/java/net/glowstone/block/blocktype/BlockType.java [hunk 2]
```diff
@@ -193,6 +195,165 @@ public final void rightClickBlock(GlowPlayer player, GlowBlock against, BlockFac
...
+     * @param isDirect Whether we are applying direct or indirect power.
+     */
+    private void traceBlockPowerStartSolid(GlowBlock srcBlock, RSManager rsManager, BlockFace toDir, boolean isDirect) {
+        // Get the destination block and ensure that it is suitable.
+        GlowBlock destBlock = srcBlock.getRelative(toDir);
+        if (destBlock == null) {
+            return;
+        }
+        Material destMat = destBlock.getType();
+        if (destMat == Material.REDSTONE_WIRE) {
```
Comment (turt2live) [inline src/main/java/net/glowstone/block/blocktype/BlockType.java]: Reduces to `if (destMat != Material.REDSTONE_WIRE) return;`

**Comment by turt2live** [comment]
Initial testing shows that this is working. However, I'd like to do more in-depth testing to make sure that all kinds of cases are working (with respect to wire & torches). Personally I'd also like to see the redstone manager cleaned up and made less... variablized.

**Comment by turt2live** [comment]
I'm personally going to hold off from passing this through for pulling until other independent people do additional testing. Redstone is a large enough feature that one person will more than likely miss over a case or forget to test something. 

Calling upon active people to test this on their own and post their results. Thus far mine look promising.

@dequis @gdude2002 @greatman 

(of course anyone else can do so too)

Edit: Please also post your test cases (even if they are duplicated by someone else). I personally tested the following:
- Redstone wire "dies" after 15 blocks. 
- 5-clock timer 
- Various logic gates
- Various contraptions

**Comment by turt2live** [comment]
@kaenganxt Additional testing has been done with the following few issues found. We weren't able to find reproduction information, but I will work to figure out how to do so.
- Blocks (such as stone) sometimes become permanent +15 sources (like redstone blocks)
- Burnout does not occur (may be out of scope if it's too complicated)
- Sometimes chunks are loaded but not ticking. We had a case where in the middle of the contraption the redstone wasn't firing.

**Comment by kaenganxt** [comment]
The latest commit should fix the problem that blocks become sources.
But i could not reproduce the case with the not tracked redstone. Any ideas?
I'm now looking into the burnout, but i think i have to add another HashMap to the RSManager. Any thoughts about that?

EDIT: @turt2live any thoughts?

**Comment by turt2live** [comment]
@kaenganxt head's up: this is no longer mergable.

When it's mergable, I'll gladly figure out how to get the infinite block bug to reappear and ensure it can't be done with these changes. As for adding another HashMap to the RS manager, I guess we have no other choice. 

As it currently stands I think that optimizing the RS manager will have to come after this PR is pulled. This is mostly due to the fact that the RS manager is complex in nature. I'll take a much more detailed look this weekend at the code base and attempt to come up with a solution that makes it less... variablized. If I can't come up with anything simple enough to be tacked on to this PR, it will have to wait. Of course others are welcome to make suggestions :)

**Comment by dequis** [comment]
Trivial conflict in ItemTable (added items - keep both sides)

Merged branch here: https://github.com/dequis/Glowstone/compare/pr-436

> head's up: this is no longer mergable.

Fuck you, now it is.

(yes, i know, i'm so evil)

**Comment by kaenganxt** [comment]
Fixed the merge conflicts.

**Comment by kaenganxt** [comment]
I tried a bit around with redstone torch burnout, but I think it's too complicated to be part of this PR. So after this gets pulled, I may again try to look into that.

**Comment by dequis** [comment]
Simple conflicts in ItemTable, keep both sides. [Branch](https://github.com/dequis/Glowstone/compare/pr-436)

**Comment by turt2live** [comment]
@SpaceManiac Confirmed that this is still valid after the rebase.

File: src/main/java/net/glowstone/RSManager.java (hunk 1 above)
Comment (socram8888) [inline src/main/java/net/glowstone/RSManager.java:29]: Couldn't this be stored on the chunk object itself?

**Comment by turt2live** [comment]
@kaenganxt After some reconsideration I don't feel as though this is ready for widespread use just yet. @coelho has brought forward some very real concerns about performance, and after some discussion I'm fairly convinced that this is not light on system resources (some quick profiling proved it's faster than vanilla, but it could use a lot more work). 

He mentioned something about octrees, which would potentially help with the speed and performance of redstone.

**Comment by Aaron1011** [comment]
@turt2live: IMHO, if it's faster than Vanilla, then performance shouldn't be a concern right now. It's already surpassing what's normally expected, so any performance improvements can come later.

**Comment by kaenganxt** [comment]
Does anyone have an idea on how the performance can be improved?

@coelho How would I use/implement your suggestion? Could you please explain a bit about octrees and how it improves the redstone manager performance.

**Comment by kaenganxt** [comment]
If someone has an idea how to improve the performance i am willing to implement it. Otherwise i think performance improvements can be done after this gets pulled since the performance is already better than in vanilla.

**Comment by turt2live** [comment]
After some more serious investigation into how this PR performs as well as an overall review of functionality, I don't think that the PR in it's current state is suitable to be pulled as-is. Ignoring the bugs (mentioned later), performance was sub-par when using a [simple world](http://i.turt2live.com/8udO.jpg) for testing. While testing I found that CPU usage was acceptable, but the RAM usage was unacceptable. The RAM usage on that world for JUST the redstone parts was 980mb, a highly unacceptable level. 

The performance concerns, as mentioned, require attention before this can really be pulled. Octrees look promising and I encourage research be done on how they work and how to write them for this particular purpose. They would allow the redstone to be limited to a smaller area instead of globally checking all devices. 

In addition to taking a look at octrees, I strongly suggest some kind of system be put in place to reduce the usage of collections as they are one of the major players in the current performance problem. Something that allows for ease of extensibility in the future for other devices, like repeaters and comparators. A component-like system (pushing the definition a bit) would likely be useful in this scenario.

There are a few known bugs with the current system. I've listed them below for consideration when writing the improved system.
- A 16-block line of powered redstone does not power a block at the end (in [this image](http://i.turt2live.com/QUVi.jpg), the torch should be off)
- Redstone blocks are not considered solid
- Timings are not perfectly vanilla (off by a reasonable amount though)

With all of the above mentioned, I've placed the PR into WIP so that the content can be worked on. Thanks!

**Comment by gdude2002** [comment]
It builds, at least: http://bamboo.gserv.me/browse/GLOW-SRV-PR436-124

**Comment by jomo** [comment]
> off by a reasonable amount though

Well, reasonable for interacting with trapdoors maybe. As soon as you use it for something that's a bit more complex, timing is very important.

---

Without looking at the vanilla code I can say that redstone is completely buggy and weird. I don't think there is a single person that fully understands how/why redstone works the way it (doesn't) work. There are a [bunch of redstone related bugs](https://bugs.mojang.com/browse/MC-72995?jql=status%20%3D%20Open%20AND%20text%20~%20%22redstone%22). We will have to reproduce these bugs, because there's always things that rely on the bugs. There's always some redstone devices that break when Mojang fixes a Redstone bug. If we don't reproduce the bugs, some devices wouldn't work in Glowstone although they do in Vanilla.

**Comment by turt2live** [comment]
@kaenganxt There has been no activity on this pull request for almost a month now. Is this still being worked on or is there some circumstance that has interfered? Please respond as soon as possible, thanks!

Additionally, this pull request is not mergeable.

**Comment by kaenganxt** [comment]
I can't seem to find a way to improve the performance with this implementation. I'm closing this PR now, but if someone has an idea, he can continue this PR. I think that we have to develop several redstone implementations and search for the one with the best performance so i may try another way of implementing it later.

**Pull request closed without merging**